### ISSUE TITLE
Proxy errors after upgrade to 2.11.1

### ISSUE THREAD

**Issue by HighWatersDev** [linkerd/linkerd2#7685]
### What is the issue?

Hey folks,
I recently upgraded from 2.10 -> 2.11.1 and seeing error messages all across the deployment. I used these guides to do tls rotation for control plane and webhooks. I'm running on AKS, using cert-manager, not using Linkerd CNI, not using network policies either.
https://linkerd.io/2.11/tasks/automatically-rotating-webhook-tls-credentials/
https://linkerd.io/2.11/tasks/automatically-rotating-control-plane-tls-credentials/

After the upgrade, I started seeing errors in proxies on both linkerd-destination and injected proxies.

### How can it be reproduced?

I initially upgraded from 2.10 -> 2.11.1 using helm. After seeing the errors, I did a clean install but the errors persisted.

### Logs, error output, etc

policy container:
```
2022-01-21T06:44:18.695936Z  INFO servers: linkerd_policy_controller_k8s_api::watch: Failed error=watch stream failed: Error reading events stream: error reading a body from connection: error reading a body from connection: Connection reset by peer (os error 104)
2022-01-21T06:44:19.697250Z  INFO servers: linkerd_policy_controller_k8s_api::watch: Restarting
2022-01-20T10:15:32.475040Z  INFO serverauthorizations: linkerd_policy_controller_k8s_api::watch: Failed error=watch stream failed: Error reading events stream: error reading a body from connection: error reading a body from connection: Connection reset by peer (os error 104)
2022-01-20T10:15:33.476285Z  INFO serverauthorizations: linkerd_policy_controller_k8s_api::watch: Restarting
```
linkerd-proxy on linkerd-destination pod:
```
[    15.798079s]  WARN ThreadId(01) policy:watch{port=8090}:controller{addr=localhost:8090}:endpoint{addr=127.0.0.1:8090}: linkerd_reconnect: Failed to connect error=Connection refused (os error 111)
[    16.299718s]  WARN ThreadId(01) policy:watch{port=8090}:controller{addr=localhost:8090}:endpoint{addr=127.0.0.1:8090}: linkerd_reconnect: Failed to connect error=Connection refused (os error 111)
```
linkerd-proxy on app pod:
```
[     1.076243s]  INFO ThreadId(02) daemon:admin{listen.addr=0.0.0.0:4191}: linkerd_app_core::serve: Connection closed error=non-HTTP connection from 127.0.0.1:36208
[     3.844111s]  INFO ThreadId(01) outbound:server{orig_dst=10.0.136.221:26379}: linkerd_proxy_api_resolve::resolve: No endpoints
[     6.843650s]  WARN ThreadId(01) outbound:server{orig_dst=10.0.136.221:26379}: linkerd_stack::failfast: TCP Logical entering failfast after 3s
[     6.843856s]  INFO ThreadId(01) outbound: linkerd_app_core::serve: Connection closed error=TCP Logical service in fail-fast
[     6.843886s]  INFO ThreadId(01) outbound: linkerd_app_core::serve: Connection closed error=TCP Logical service in fail-fast
```
another pod:
```
[     1.049246s]  INFO ThreadId(02) daemon:admin{listen.addr=0.0.0.0:4191}: linkerd_app_core::serve: Connection closed error=non-HTTP connection from 127.0.0.1:45840
[   139.764538s]  WARN ThreadId(01) outbound:server{orig_dst=10.0.0.1:443}:controller{addr=linkerd-dst-headless.linkerd.svc.cluster.local:8086}:endpoint{addr=10.4.17.58:8086}: linkerd_reconnect: Service failed error=channel closed
[   140.864223s]  WARN ThreadId(01) policy:watch{port=10254}:controller{addr=linkerd-policy.linkerd.svc.cluster.local:8090}:endpoint{addr=10.4.17.58:8090}: linkerd_reconnect: Service failed error=channel closed
```

### output of `linkerd check -o short`

```text
Linkerd core checks
===================

linkerd-identity
----------------
‼ issuer cert is valid for at least 60 days
    issuer certificate will expire on 2022-01-25T22:47:53Z
    see https://linkerd.io/2.11/checks/#l5d-identity-issuer-cert-not-expiring-soon for hints

Status check results are √

Linkerd extensions checks
=========================


Status check results are √
```

### Environment

- Kubernetes version: 1.20.9
- Azure AKS
- Linkerd version: 2.11.1

**Comment by olix0r** [comment]
For some reason, the policy controller is unable to connect to the Kubernetes API. We've seen other weird behavior in AKS, but not this...

Can you try install/upgrade with `--set policyController.image.version=edge-22.1.3`. We made changes for #7275 that may fix TLS negotiation issues in some clusters.

**Comment by olix0r** [comment]
Additionally, using `--set policyController.logLevel="debug"` may help to shine light on the nature of the connection issue.

**Comment by HighWatersDev** [comment]
Thanks for the quick follow-up. Here's the log file from policy container with debug on. After I upgraded the image as you suggested:
[linkerd-destination.log](https://github.com/linkerd/linkerd2/files/7928230/linkerd-destination.log)
 I also see similar logs on other proxy containers but it only seems to happen once:
 ```text
[     1.078321s]  INFO ThreadId(02) daemon:admin{listen.addr=0.0.0.0:4191}: linkerd_app_core::serve: Connection closed error=non-HTTP connection from 127.0.0.1:38290
```
```text
[   294.794607s]  INFO ThreadId(02) daemon:admin{listen.addr=0.0.0.0:4191}: linkerd_app_core::serve: Connection closed error=TLS detection timed out
[   594.743469s]  INFO ThreadId(02) daemon:admin{listen.addr=0.0.0.0:4191}: linkerd_app_core::serve: Connection closed error=TLS detection timed out
```

**Comment by olix0r** [comment]
Thanks. I now realize that the `policy-controller`'s `INFO` messages are innocuous -- they're just indicating the normal stream expiration/reconnect that occurs with the API client.

The logs like:
```
[     1.076243s]  INFO ThreadId(02) daemon:admin{listen.addr=0.0.0.0:4191}: linkerd_app_core::serve: Connection closed error=non-HTTP connection from 127.0.0.1:36208
```

indicate that some connection to the proxy's admin server from localhost is failing. This could be the new `postStartHook` hitting the proxy's /ready endpoint; or this could be some other pod-local traffic? I'm not sure why you would see this--perhaps due to CPU resource constraints/contention? _Something_ is connecting the proxy's admin port and isn't sending any data on the connection. If the pod enters the `Running` state, though, I'd ignore it for now.

You said that an app pod shows:

```
[     3.844111s]  INFO ThreadId(01) outbound:server{orig_dst=10.0.136.221:26379}: linkerd_proxy_api_resolve::resolve: No endpoints
[     6.843650s]  WARN ThreadId(01) outbound:server{orig_dst=10.0.136.221:26379}: linkerd_stack::failfast: TCP Logical entering failfast after 3s
[     6.843856s]  INFO ThreadId(01) outbound: linkerd_app_core::serve: Connection closed error=TCP Logical service in fail-fast
[     6.843886s]  INFO ThreadId(01) outbound: linkerd_app_core::serve: Connection closed error=TCP Logical service in fail-fast
```

Is `10.0.136.221` the IP of a kubernetes service? Does the service have endpoints? I'd try using `linkerd diagnostics endpoints <service-fqdn>`--I'd expect that to return no endpoints--and comparing that to `kubectl get endpoints -n $ns $service`. Does kubernetes find endpoints for this service?

We'll want to confirm that the target service is actually running and healthy. If the proxy is running on the server-side pods, is there anything in the logs?

**Comment by HighWatersDev** [comment]
Thank you for the info. Even though it's hard, but I'll ignore those other messages for now lol.
As far as errors `no endpoint`, it only seems to happen when I first create the pod and it doesn't repeat after that. When I checked, both commands `linkerd diagnostics` and `kubectl get endpoints` return the actual endpoints that match the ip address in the error message.
